File: android/src/main/kotlin/com/example/qr_scanner_master/BarcodeScanner.kt
package com.example.qr_scanner_master

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Point
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.barcode.common.Barcode
import com.google.mlkit.vision.common.InputImage
import java.io.File

class BarcodeScanner(private val context: Context) {
    
    private val scanner = BarcodeScanning.getClient()
    
    fun scanFromImagePath(imagePath: String, options: ScanOptions, callback: (List<ScanResult>) -> Unit) {
        try {
            val file = File(imagePath)
            if (!file.exists()) {
                callback(emptyList())
                return
            }
            
            val bitmap = BitmapFactory.decodeFile(imagePath)
            if (bitmap == null) {
                callback(emptyList())
                return
            }
            
            scanFromBitmap(bitmap, options, callback)
        } catch (e: Exception) {
            callback(emptyList())
        }
    }
    
    fun scanFromBytes(imageBytes: ByteArray, options: ScanOptions, callback: (List<ScanResult>) -> Unit) {
        try {
            val bitmap = BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.size)
            if (bitmap == null) {
                callback(emptyList())
                return
            }
            
            scanFromBitmap(bitmap, options, callback)
        } catch (e: Exception) {
            callback(emptyList())
        }
    }
    
    private fun scanFromBitmap(bitmap: Bitmap, options: ScanOptions, callback: (List<ScanResult>) -> Unit) {
        val image = InputImage.fromBitmap(bitmap, 0)
        
        scanner.process(image)
            .addOnSuccessListener { barcodes ->
                val results = barcodes.mapNotNull { barcode ->
                    if (shouldIncludeBarcode(barcode, options)) {
                        convertToScanResult(barcode)
                    } else {
                        null
                    }
                }
                callback(results)
            }
            .addOnFailureListener {
                callback(emptyList())
            }
    }
    
    private fun shouldIncludeBarcode(barcode: Barcode, options: ScanOptions): Boolean {
        if (options.formats.isEmpty()) {
            return true // Include all formats if none specified
        }
        
        val formatName = getBarcodeFormatName(barcode.format)
        return options.formats.contains(formatName)
    }
    
    private fun getBarcodeFormatName(format: Int): String {
        return when (format) {
            Barcode.FORMAT_QR_CODE -> "QR_CODE"
            Barcode.FORMAT_EAN_8 -> "EAN_8"
            Barcode.FORMAT_EAN_13 -> "EAN_13"
            Barcode.FORMAT_CODE_39 -> "CODE_39"
            Barcode.FORMAT_CODE_93 -> "CODE_93"
            Barcode.FORMAT_CODE_128 -> "CODE_128"
            Barcode.FORMAT_CODABAR -> "CODABAR"
            Barcode.FORMAT_ITF -> "ITF"
            Barcode.FORMAT_UPC_A -> "UPC_A"
            Barcode.FORMAT_UPC_E -> "UPC_E"
            Barcode.FORMAT_DATA_MATRIX -> "DATA_MATRIX"
            Barcode.FORMAT_AZTEC -> "AZTEC"
            Barcode.FORMAT_PDF417 -> "PDF_417"
            else -> "UNKNOWN"
        }
    }
    
    private fun convertToScanResult(barcode: Barcode): ScanResult {
        val cornerPoints = barcode.cornerPoints?.map { point ->
            ScanPoint(point.x.toDouble(), point.y.toDouble())
        } ?: emptyList()
        
        return ScanResult(
            data = barcode.rawValue ?: "",
            format = getBarcodeFormatName(barcode.format),
            timestamp = System.currentTimeMillis(),
            cornerPoints = cornerPoints,
            metadata = createMetadata(barcode)
        )
    }
    
    private fun createMetadata(barcode: Barcode): Map<String, Any> {
        val metadata = mutableMapOf<String, Any>()
        
        // Add barcode-specific data
        barcode.boundingBox?.let { rect ->
            metadata["boundingBox"] = mapOf(
                "left" to rect.left,
                "top" to rect.top,
                "right" to rect.right,
                "bottom" to rect.bottom
            )
        }
        
        // Add format-specific data
        when (barcode.valueType) {
            Barcode.TYPE_URL -> {
                barcode.url?.let { url ->
                    metadata["url"] = mapOf(
                        "title" to (url.title ?: ""),
                        "url" to (url.url ?: "")
                    )
                }
            }
            Barcode.TYPE_EMAIL -> {
                barcode.email?.let { email ->
                    metadata["email"] = mapOf(
                        "address" to (email.address ?: ""),
                        "subject" to (email.subject ?: ""),
                        "body" to (email.body ?: "")
                    )
                }
            }
            Barcode.TYPE_PHONE -> {
                barcode.phone?.let { phone ->
                    metadata["phone"] = mapOf(
                        "number" to (phone.number ?: ""),
                        "type" to phone.type
                    )
                }
            }
            Barcode.TYPE_SMS -> {
                barcode.sms?.let { sms ->
                    metadata["sms"] = mapOf(
                        "phoneNumber" to (sms.phoneNumber ?: ""),
                        "message" to (sms.message ?: "")
                    )
                }
            }
            Barcode.TYPE_WIFI -> {
                barcode.wifi?.let { wifi ->
                    metadata["wifi"] = mapOf(
                        "ssid" to (wifi.ssid ?: ""),
                        "password" to (wifi.password ?: ""),
                        "encryptionType" to wifi.encryptionType
                    )
                }
            }
            Barcode.TYPE_GEO -> {
                barcode.geoPoint?.let { geo ->
                    metadata["geoPoint"] = mapOf(
                        "latitude" to geo.lat,
                        "longitude" to geo.lng
                    )
                }
            }
            Barcode.TYPE_CONTACT_INFO -> {
                barcode.contactInfo?.let { contact ->
                    metadata["contactInfo"] = mapOf(
                        "name" to mapOf(
                            "first" to (contact.name?.first ?: ""),
                            "last" to (contact.name?.last ?: ""),
                            "middle" to (contact.name?.middle ?: ""),
                            "prefix" to (contact.name?.prefix ?: ""),
                            "suffix" to (contact.name?.suffix ?: ""),
                            "formattedName" to (contact.name?.formattedName ?: "")
                        ),
                        "organization" to (contact.organization ?: ""),
                        "title" to (contact.title ?: ""),
                        "phones" to (contact.phones?.map { phone ->
                            mapOf(
                                "number" to (phone.number ?: ""),
                                "type" to phone.type
                            )
                        } ?: emptyList()),
                        "emails" to (contact.emails?.map { email ->
                            mapOf(
                                "address" to (email.address ?: ""),
                                "type" to email.type
                            )
                        } ?: emptyList()),
                        "urls" to (contact.urls ?: emptyList()),
                        "addresses" to (contact.addresses?.map { address ->
                            mapOf(
                                "addressLines" to (address.addressLines?.toList() ?: emptyList()),
                                "type" to address.type
                            )
                        } ?: emptyList())
                    )
                }
            }
            Barcode.TYPE_CALENDAR_EVENT -> {
                barcode.calendarEvent?.let { event ->
                    metadata["calendarEvent"] = mapOf(
                        "summary" to (event.summary ?: ""),
                        "description" to (event.description ?: ""),
                        "location" to (event.location ?: ""),
                        "organizer" to (event.organizer ?: ""),
                        "status" to (event.status ?: ""),
                        "start" to (event.start?.rawValue ?: ""),
                        "end" to (event.end?.rawValue ?: "")
                    )
                }
            }
            Barcode.TYPE_DRIVER_LICENSE -> {
                barcode.driverLicense?.let { license ->
                    metadata["driverLicense"] = mapOf(
                        "documentType" to (license.documentType ?: ""),
                        "firstName" to (license.firstName ?: ""),
                        "middleName" to (license.middleName ?: ""),
                        "lastName" to (license.lastName ?: ""),
                        "gender" to (license.gender ?: ""),
                        "addressStreet" to (license.addressStreet ?: ""),
                        "addressCity" to (license.addressCity ?: ""),
                        "addressState" to (license.addressState ?: ""),
                        "addressZip" to (license.addressZip ?: ""),
                        "licenseNumber" to (license.licenseNumber ?: ""),
                        "issueDate" to (license.issueDate ?: ""),
                        "expiryDate" to (license.expiryDate ?: ""),
                        "birthDate" to (license.birthDate ?: ""),
                        "issuingCountry" to (license.issuingCountry ?: "")
                    )
                }
            }
        }
        
        return metadata
    }
}

data class ScanOptions(
    val formats: List<String> = emptyList(),
    val enableFlash: Boolean = false,
    val autoFocus: Boolean = true,
    val multiScan: Boolean = false,
    val maxScans: Int = 1,
    val beepOnScan: Boolean = true,
    val vibrateOnScan: Boolean = true,
    val showOverlay: Boolean = true,
    val overlayColor: Int = 0xFF00FF00.toInt(),
    val restrictScanArea: Boolean = false,
    val scanAreaRatio: Float = 0.7f,
    val timeoutSeconds: Int = 0,
    val returnImage: Boolean = false,
    val imageQuality: Float = 0.8f,
    val detectInverted: Boolean = false,
    val cameraResolution: String = "MEDIUM",
    val cameraFacing: String = "BACK"
) {
    companion object {
        fun fromMap(map: Map<String, Any>): ScanOptions {
            val formatsList = map["formats"] as? List<*>
            return ScanOptions(
                formats = formatsList?.filterIsInstance<String>() ?: emptyList(),
                enableFlash = map["enableFlash"] as? Boolean ?: false,
                autoFocus = map["autoFocus"] as? Boolean ?: true,
                multiScan = map["multiScan"] as? Boolean ?: false,
                maxScans = (map["maxScans"] as? Number)?.toInt() ?: 1,
                beepOnScan = map["beepOnScan"] as? Boolean ?: true,
                vibrateOnScan = map["vibrateOnScan"] as? Boolean ?: true,
                showOverlay = map["showOverlay"] as? Boolean ?: true,
                overlayColor = (map["overlayColor"] as? Number)?.toInt() ?: 0xFF00FF00.toInt(),
                restrictScanArea = map["restrictScanArea"] as? Boolean ?: false,
                scanAreaRatio = (map["scanAreaRatio"] as? Number)?.toFloat() ?: 0.7f,
                timeoutSeconds = (map["timeoutSeconds"] as? Number)?.toInt() ?: 0,
                returnImage = map["returnImage"] as? Boolean ?: false,
                imageQuality = (map["imageQuality"] as? Number)?.toFloat() ?: 0.8f,
                detectInverted = map["detectInverted"] as? Boolean ?: false,
                cameraResolution = map["cameraResolution"] as? String ?: "MEDIUM",
                cameraFacing = map["cameraFacing"] as? String ?: "BACK"
            )
        }
    }
}

data class ScanResult(
    val data: String,
    val format: String,
    val timestamp: Long,
    val cornerPoints: List<ScanPoint>,
    val metadata: Map<String, Any>
) {
    fun toMap(): Map<String, Any> {
        return mapOf(
            "data" to data,
            "format" to format,
            "timestamp" to timestamp,
            "cornerPoints" to cornerPoints.map { it.toMap() },
            "metadata" to metadata
        )
    }
}

data class ScanPoint(
    val x: Double,
    val y: Double
) {
    fun toMap(): Map<String, Any> {
        return mapOf(
            "x" to x,
            "y" to y
        )
    }
}

File: ios/Classes/BarcodeScanner.swift
import UIKit
import Vision
import AVFoundation

class BarcodeScanner {
    
    func scanFromImagePath(imagePath: String, options: ScanOptions, completion: @escaping ([ScanResult]) -> Void) {
        guard let image = UIImage(contentsOfFile: imagePath) else {
            completion([])
            return
        }
        
        scanFromImage(image: image, options: options, completion: completion)
    }
    
    func scanFromBytes(imageData: Data, options: ScanOptions, completion: @escaping ([ScanResult]) -> Void) {
        guard let image = UIImage(data: imageData) else {
            completion([])
            return
        }
        
        scanFromImage(image: image, options: options, completion: completion)
    }
    
    private func scanFromImage(image: UIImage, options: ScanOptions, completion: @escaping ([ScanResult]) -> Void) {
        guard let cgImage = image.cgImage else {
            completion([])
            return
        }
        
        let request = VNDetectBarcodesRequest { request, error in
            guard error == nil else {
                completion([])
                return
            }
            
            let results = request.results as? [VNBarcodeObservation] ?? []
            let scanResults = results.compactMap { observation -> ScanResult? in
                return self.convertToScanResult(observation: observation, options: options)
            }
            
            completion(scanResults)
        }
        
        // Configure supported symbologies based on options
        if !options.formats.isEmpty {
            request.symbologies = options.formats.compactMap { formatName in
                self.getVisionSymbology(from: formatName)
            }
        }
        
        let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
        
        do {
            try handler.perform([request])
        } catch {
            completion([])
        }
    }
    
    private func convertToScanResult(observation: VNBarcodeObservation, options: ScanOptions) -> ScanResult? {
        guard let payloadString = observation.payloadStringValue else { return nil }
        
        let formatName = getFormatName(from: observation.symbology)
        
        // Check if format should be included
        if !options.formats.isEmpty && !options.formats.contains(formatName) {
            return nil
        }
        
        // Convert corner points
        let cornerPoints = [
            observation.topLeft,
            observation.topRight,
            observation.bottomRight,
            observation.bottomLeft
        ].map { point in
            ScanPoint(x: Double(point.x), y: Double(point.y))
        }
        
        // Create metadata
        let metadata = createMetadata(from: observation)
        
        return ScanResult(
            data: payloadString,
            format: formatName,
            timestamp: Int64(Date().timeIntervalSince1970 * 1000),
            cornerPoints: cornerPoints,
            metadata: metadata
        )
    }
    
    private func getVisionSymbology(from formatName: String) -> VNBarcodeSymbology? {
        switch formatName {
        case "QR_CODE":
            return .qr
        case "EAN_8":
            return .ean8
        case "EAN_13":
            return .ean13
        case "CODE_39":
            return .code39
        case "CODE_93":
            return .code93
        case "CODE_128":
            return .code128
        case "CODABAR":
            return .codabar
        case "ITF":
            return .itf14
        case "UPC_A":
            return .upce
        case "UPC_E":
            return .upce
        case "DATA_MATRIX":
            return .dataMatrix
        case "AZTEC":
            return .aztec
        case "PDF_417":
            return .pdf417
        default:
            return nil
        }
    }
    
    private func getFormatName(from symbology: VNBarcodeSymbology) -> String {
        switch symbology {
        case .qr:
            return "QR_CODE"
        case .ean8:
            return "EAN_8"
        case .ean13:
            return "EAN_13"
        case .code39:
            return "CODE_39"
        case .code93:
            return "CODE_93"
        case .code128:
            return "CODE_128"
        case .codabar:
            return "CODABAR"
        case .itf14:
            return "ITF"
        case .upce:
            return "UPC_E"
        case .dataMatrix:
            return "DATA_MATRIX"
        case .aztec:
            return "AZTEC"
        case .pdf417:
            return "PDF_417"
        default:
            return "UNKNOWN"
        }
    }
    
    private func createMetadata(from observation: VNBarcodeObservation) -> [String: Any] {
        var metadata: [String: Any] = [:]
        
        // Add bounding box
        let boundingBox = observation.boundingBox
        metadata["boundingBox"] = [
            "x": boundingBox.origin.x,
            "y": boundingBox.origin.y,
            "width": boundingBox.size.width,
            "height": boundingBox.size.height
        ]
        
        // Add confidence
        metadata["confidence"] = observation.confidence
        
        // Add format-specific data based on descriptor
        if let descriptor = observation.barcodeDescriptor {
            switch descriptor {
            case let qrDescriptor as CIQRCodeDescriptor:
                metadata["qrCode"] = [
                    "errorCorrectionLevel": qrDescriptor.errorCorrectionLevel.rawValue,
                    "symbolVersion": qrDescriptor.symbolVersion,
                    "maskPattern": qrDescriptor.maskPattern
                ]
                
            case let aztecDescriptor as CIAztecCodeDescriptor:
                metadata["aztecCode"] = [
                    "isCompact": aztecDescriptor.isCompact,
                    "layerCount": aztecDescriptor.layerCount,
                    "dataCodewordCount": aztecDescriptor.dataCodewordCount
                ]
                
            case let pdf417Descriptor as CIPDF417CodeDescriptor:
                metadata["pdf417"] = [
                    "isCompact": pdf417Descriptor.isCompact,
                    "rowCount": pdf417Descriptor.rowCount,
                    "columnCount": pdf417Descriptor.columnCount
                ]
                
            case let dataMatrixDescriptor as CIDataMatrixCodeDescriptor:
                metadata["dataMatrix"] = [
                    "rowCount": dataMatrixDescriptor.rowCount,
                    "columnCount": dataMatrixDescriptor.columnCount,
                    "eccVersion": dataMatrixDescriptor.eccVersion.rawValue
                ]
                
            default:
                break
            }
        }
        
        return metadata
    }
}

struct ScanOptions {
    let formats: [String]
    let enableFlash: Bool
    let autoFocus: Bool
    let multiScan: Bool
    let maxScans: Int
    let beepOnScan: Bool
    let vibrateOnScan: Bool
    let showOverlay: Bool
    let overlayColor: Int
    let restrictScanArea: Bool
    let scanAreaRatio: CGFloat
    let timeoutSeconds: Int
    let returnImage: Bool
    let imageQuality: CGFloat
    let detectInverted: Bool
    let cameraResolution: String
    let cameraFacing: String
    
    static func fromDictionary(_ dict: [String: Any]) -> ScanOptions {
        return ScanOptions(
            formats: dict["formats"] as? [String] ?? [],
            enableFlash: dict["enableFlash"] as? Bool ?? false,
            autoFocus: dict["autoFocus"] as? Bool ?? true,
            multiScan: dict["multiScan"] as? Bool ?? false,
            maxScans: dict["maxScans"] as? Int ?? 1,
            beepOnScan: dict["beepOnScan"] as? Bool ?? true,
            vibrateOnScan: dict["vibrateOnScan"] as? Bool ?? true,
            showOverlay: dict["showOverlay"] as? Bool ?? true,
            overlayColor: dict["overlayColor"] as? Int ?? 0xFF00FF00,
            restrictScanArea: dict["restrictScanArea"] as? Bool ?? false,
            scanAreaRatio: CGFloat(dict["scanAreaRatio"] as? Double ?? 0.7),
            timeoutSeconds: dict["timeoutSeconds"] as? Int ?? 0,
            returnImage: dict["returnImage"] as? Bool ?? false,
            imageQuality: CGFloat(dict["imageQuality"] as? Double ?? 0.8),
            detectInverted: dict["detectInverted"] as? Bool ?? false,
            cameraResolution: dict["cameraResolution"] as? String ?? "MEDIUM",
            cameraFacing: dict["cameraFacing"] as? String ?? "BACK"
        )
    }
}

struct ScanResult {
    let data: String
    let format: String
    let timestamp: Int64
    let cornerPoints: [ScanPoint]
    let metadata: [String: Any]
    
    func toDictionary() -> [String: Any] {
        return [
            "data": data,
            "format": format,
            "timestamp": timestamp,
            "cornerPoints": cornerPoints.map { $0.toDictionary() },
            "metadata": metadata
        ]
    }
}

struct ScanPoint {
    let x: Double
    let y: Double
    
    func toDictionary() -> [String: Any] {
        return [
            "x": x,
            "y": y
        ]
    }
}
